### ISSUE TITLE
How can I reduce the false-positive mutants for strings?

### ISSUE THREAD

**Issue by MartinThoma** [boxed/mutmut#175]
I just ran `mutmut` over my toy project [mpu](https://github.com/MartinThoma/mpu) and got a lot of false positives for strings.

## False Positive String Mutations

**Exception messages**: They should never be mutated; all of the mutants were false-positives like this:

```
-        raise ValueError("All primes are prime factors of 0.")
+        raise ValueError("XXAll primes are prime factors of 0.XX")
```

**print values**: I neither

```
-    print("Number of datapoints: {datapoints}".format(datapoints=len(df)))
+    print("XXNumber of datapoints: {datapoints}XX".format(datapoints=len(df)))
```

## Valuable String Mutations

The following examples could potentially show issues:

**comparisons**:

```
-            if column_type == "other":
+            if column_type == "XXotherXX":
```

**assignments**: I see that assignments are hard to tell. The following could either be in a comparison or in a print (it is a print):

```
     table = [
-        ["Column name", "Non-nan", "mean", "std", "min", "25%", "50%", "75%", "max"]
+        ["Column name", "XXNon-nanXX", "mean", "std", "min", "25%", "50%", "75%", "max"]
     ]
```

## Solutions

Is it possible to configure mutmut to completely disable string mutations or to disable it for those cases?

I have seen [Advanced whitelisting and configuraton](https://github.com/boxed/mutmut/#advanced-whitelisting-and-configuration), but I don't understand how to apply it in my case.

**Comment by boxed** [comment]
> Exception messages: They should never be mutated; all of the mutants were false-positives like this:

I _strongly_ disagree with this assertion. Double digit percentages of the bugs we've found at work with mutation testing has been in the exception messages. These were all real bugs. 

> print values

Why is this a false positive? Either the print is an important part of the program and you should test it, or it is just accidentally left over print debugging and should be cleaned up. What is the third option?

> Is it possible to configure mutmut to completely disable string mutations or to disable it for those cases?

Yes. It can be as simple as adding `mutmut_config.py` with this context:

```python
def pre_mutation(context):
    if current_source_line.strip().startswith('print('):
        context.skip = True
```

**Comment by MartinThoma** [comment]
> bugs we've found at work with mutation testing has been in the exception messages

Ok. Maybe offtopic here, but I would be curious how such a bug would look like :-)

**Comment by MartinThoma** [comment]
> Why is this a false positive? Either the print is an important part of the program and you should test it, or it is just accidentally left over print debugging and should be cleaned up. What is the third option?

I don't see the point of testing the output for the user. Yes, the string is important. It gives the user information. But what is the point of copying exactly the same string to a unit test?

## Minimal Example

The following `factorize.py` is an application which lets the user input an integer and returns its factors:

```python
# Core Library modules
from typing import List

# Third party modules
import click


@click.command()
@click.argument("number", type=int, required=True)
def cli(number):
    if number == 0:
        print("Zero cannot be factorized")
    else:
        factors = factorize(number)
        print(f"{number} = {' x '.join([str(f) for f in factors])}")


def factorize(number: int) -> List[int]:
    if number == 0:
        raise ValueError("Zero cannot be factorized.")
    factors = []
    if number < 0:
        factors.append(-1)
        number *= -1
    test = 2
    while number > 1:
        while number % test == 0:
            factors.append(test)
            number = number // test
        test += 1
    return factors


if __name__ == "__main__":
    cli()
```

## Testing MVCE

To test it, I would write the `test_factorize.py` as follows and run it with pytest:

```python
# Third party modules
import hypothesis.strategies as s
import pytest
from click.testing import CliRunner
from hypothesis import given

# First party modules
from factorize import cli, factorize


@given(s.integers(min_value=-(10 ** 6), max_value=10 ** 6))
def test_factorize(number):
    if number == 0:
        with pytest.raises(ValueError):
            factorize(number)
    else:
        factors = factorize(number)
        product = 1
        for factor in factors:
            product *= factor
        assert product == number


def test_factorize_12():
    assert factorize(12) == [2, 2, 3]


def test_factorize_minus_12():
    assert factorize(-12) == [-1, 2, 2, 3]


def test_sync():
    runner = CliRunner()
    result = runner.invoke(cli, ["12"])
    assert result.exit_code == 0
    assert "2 x 2 x 3" in result.output
```

## mutmut on MVCE
I ran it with `mutmut run --paths-to-mutate factorize.py` and got:

```
⠸ 35/35  🎉 20  ⏰ 7  🤔 0  🙁 8  🔇 0
```

and the HTML output, ignoring the timeout mutations because those tests result in an infinite loop which means the test suite would fail:

```
Mutant 7:
-        print("Zero cannot be factorized")
+        print("XXZero cannot be factorizedXX")

Mutant 12:
-        raise ValueError("Zero cannot be factorized.")
+        raise ValueError("XXZero cannot be factorized.XX")
```

The other 6 mutants with a sad smiley are understandable, although I don't think the tests would require a change.

Would you add / adjust the tests? Do I miss maybe a whole category of potential bugs?

**Comment by boxed** [comment]
>>    bugs we've found at work with mutation testing has been in the exception messages

> Ok. Maybe offtopic here, but I would be curious how such a bug would look like :-)

We almost exclusively mutation test basic libraries and there error message are super important. I often say "the true measure of a person is in crisis, and the true measure of a program is in error messages" :P

**Comment by boxed** [comment]
Just looking at it quickly I would do this:

```python
def test_sync():
    runner = CliRunner()
    result = runner.invoke(cli, ["12"])
    assert result.exit_code == 0
    assert "2 x 2 x 3" == result.output.strip()

def test_sync_0():
    runner = CliRunner()
    result = runner.invoke(cli, ["0"])
    assert result.exit_code == 0
    assert "Zero cannot be factorized" == result.output.strip()
```

I believe this is a better test suite and will kill some mutants.

**Comment by blueskyjunkie** [comment]
For the exceptions you should also have tests like this (example uses pytest, you can definitely do equivalent using unittest, and presumably other frameworks):

```python
import pytest

def test_raises():
    with pytest.raises(ValueError, match=r"^Zero cannot be factorized"):
        result = factorize(0)
```

This tests both that the correct type of exception is raised when you expect it, and that the exception raised has the correct message since you might use the same exception multiple times for different error messages.

**Comment by MartinThoma** [comment]
Cool, I didn't know the match parameter - thank you!

However, I'm uncertain if I really like to use it. I'm thinking about creating a more customized Exception which inherits from ValueError - something like ZeroFactorizationError. Then the message is something for making it more clear, but definitely not something I would check. I don't see the value in having the same string in both, the test and the code. Which kind of bug would this catch?

**Comment by boxed** [comment]
In our experience in building iommi, it catches the most common bugs: bad error messages :P

If you're developing a library this may be the most important part of the entire library, while if you're developing a web app where the error is just logged to sentry or something, well then it doesn't matter very much. Mutmut doesn't know what you're developing...

**Comment by blueskyjunkie** [comment]
@MartinThoma I disagree that you would "never check this because I've designed/implemented to not need the test". The problem with software development is often not how it is implemented now, but how it changes over time. What if someone adds a new exception in future in the `factorize` function and reuses the type with a different error message? You want tests to clearly indicate that a new logical error has been introduced by the change (if indeed that is the case). Yes, you could document somewhere that the exception type is not supposed to be reused, but what if that person forgets, doesn't notice, is careless, whatever; you still want the tests to clearly indicate that a problem has been introduced. Software development has to account for present and future human cognitive frailty as well correctness in delivering business value.

Think of unit/integration tests as the definition of correctness of your software _at this moment in time_. As the software evolves, the tests should evolve with the software to continue to define correctness. Historically test _coverage_ has been used as a proxy for test quality in absence of other metrics (and it does help), however `mutmut` introduces the possibility of actually testing the quality of the tests.

**Comment by MartinThoma** [comment]
> What if someone adds a new exception in future in the factorize function and reuses the type with a different error message?

Good point, I didn't think of that :+1: - So what we test for is actually a combination:

1. Another usage of the same Exception is introduced (although it probably shouldn't)
2. In any of the two cases, the wrong one is thrown

I now understand the point with Exception messages and agree with it. Do you have a similar example for log messages / output printed for the end user?

**Comment by MartinThoma** [comment]
> Historically test coverage has been used as a proxy for test quality in absence of other metrics (and it does help), however mutmut introduces the possibility of actually testing the quality of the tests.

By the way, is it possible to get a line coverage adjusted by the mutation tests? I imagine it to work like this:

1. total lines are calculated however it is currently done with pytest-cov
2. lines which are not covered according to pytest-cov are counted as "not covered"
3. lines in which at least one mutant survives are counted as "not covered"

To me, this would also be a more valuable output of mutmut: Just one output per line instead of potentially several dozens. (I can open another issue for that discussion if you want :-) )

**Comment by MartinThoma** [comment]
Thank you both for taking so much time to help me <3

**Comment by boxed** [comment]
There's a philosophical difference of what mutation testing is for between different tools. I think the report and the numbers are mostly irrelevant. I've even seen people use the mutation score for arguably political leverage which I find a bit icky. 

To me it's a tool to find holes in the test suite that you want to work with. Then the report "x mutants killed, y survived" is enough.

**Comment by toonarmycaptain** [comment]
If you follow a pattern to maximise DRY and minimise unnecessary code churn, you might doing `from script import bad_error_msg, not_so_bad_error_msg` and then testing that those messages are thrown as expected (you don't really need to test that python correctly passes/prints a given string). When `mutmut` mutates that string...well you're testing that the mutated string is passed, which it will be as expected, and then you're down a rabbit hole. If lots of tests might throw the same error message...that's a lot of code churn just to fix a spelling or phrasing issue if you have to change it in every test. 

I don't write this as a concrete criticism, just making the point. I don't know how to avoid the issue without a lot of introspecting code and tests to see where expected values in tests are pulled from the code itself, and that sounds like it would be either rather error-prone or too broad.

**Comment by boxed** [comment]
You should only assert the error message in one test. That's one test to change for each error message changed. We do this in several libs. 

If you don't care so much about the error messages you can whitelist them either with the pragma or (maybe more reasonably) by whitelisting all lines containing "raise " with the advanced whitelisting system.

**Comment by DudeNr33** [comment]
I'd like to come back to one portion of the original question:

> Is it possible to configure mutmut to completely disable string mutations [...]?

While I agree on why all discussed cases can be important in some contexts, I think having the option to select what types of mutation you want to test could be really beneficial if you try to introduce mutation testing in an existing, larger project.
Looking into the source code I noticed the following dictionary:

```python
mutations_by_type = {
    'operator': dict(value=operator_mutation),
    'keyword': dict(value=keyword_mutation),
    'number': dict(value=number_mutation),
    'name': dict(value=name_mutation),
    'string': dict(value=string_mutation),
    'argument': dict(children=argument_mutation),
    'or_test': dict(children=and_or_test_mutation),
    'and_test': dict(children=and_or_test_mutation),
    'lambdef': dict(children=lambda_mutation),
    'expr_stmt': dict(children=expression_mutation),
    'decorator': dict(children=decorator_mutation),
    'annassign': dict(children=expression_mutation),
}
```

Being able to run something like:
```shell
mutmut run --disable-mutation-type=string
```
or
```shell
mutmut run --enable-mutation-type=operator,number
```
might be a good way to control what types of mutations are run. 
This way the user could decide which mutation types are the most important to him. Maybe I am developing an application where some measurements are frequently checked against some thresholds, then I might decide that checking for potential bugs in comparisons of numbers is the highest priority, and only after getting rid of those mutants I want to check for other things as well.

Or is this maybe already possible through ``mutmut_config.py``? 
The given example only covers how to skip tests if they contain a certain string, but maybe it would be possible to check what type of transformation is going to be performed?

**Comment by boxed** [comment]
This would be a great feature and something I've even discussed with other mutation testers for python. But I just haven't had a need for it myself so that's why it'snot implementated. A PR is very welcome.

**Comment by DudeNr33** [comment]
I could give it a try. One question before I start: is it enough to add a new command line option and skip over the undesired mutation types, or is there anything else that needs to be considered?

**Comment by boxed** [comment]
There should probably be an exclude and an include type flag. And they should take a list of things. The existing options are read from the config file automatically so if you follow that pattern you'll get that feature too.

**Comment by MartinThoma** [comment]
Two options I like would be

* `--select=operator,number`: Only execute operator and number mutations ([as in flake8](https://flake8.pycqa.org/en/latest/user/options.html#cmdoption-flake8-select))
* `--exclude=operator,number`: Run all mutations except for operator and number mutations. `--disable` would sound fine to me as well.

Keywords I've considered:

* ignore: Sounds a bit as if it's still executed, but not reported (e.g. as it happens in flake8)
* include/enable: Sounds like adding extras to another set, e.g. as if there are some enabled by default and some need to be included.

**Comment by DudeNr33** [comment]
Ah, I forgot to link the PR (#225). 
The next ``mutmut`` version will provide two new options:
* ``--enable-mutation-types``
* ``--disable-mutation-types``

See the "advanced configuration and whitelisting" section in the readme.
